packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -84,6 +84,49 @@ const Duration _kCursorBlinkHalfPeriod = Duration(milliseconds: 500);
 // is shown in an obscured text field.
 const int _kObscureShowLatestCharCursorTicks = 3;
 
+class _RenderCompositionCallback extends RenderProxyBox {
+  _RenderCompositionCallback(this.compositeCallback);
+
+  final void Function(Layer) compositeCallback;
+
+  bool get enabled => _enabled;
+  bool _enabled = false;
+  VoidCallback? _cancelCallback;
+  set enabled(bool newValue) {
+    if (!newValue) {
+      _cancelCallback?.call();
+      _cancelCallback = null;
+    }
+    _enabled = newValue;
+  }
+
+  @override
+  void paint(PaintingContext context, ui.Offset offset) {
+    if (enabled) {
+      _cancelCallback ??= context.addCompositionCallback(compositeCallback);
+    }
+    super.paint(context, offset);
+  }
+}
+
+class _CompositionCallback extends SingleChildRenderObjectWidget {
+  const _CompositionCallback({ required this.compositeCallback, this.enabled = false, super.child });
+  final void Function(Layer) compositeCallback;
+  final bool enabled;
+
+  @override
+  RenderObject createRenderObject(BuildContext context) {
+    assert(!enabled);
```

Comment:
This seems kinda odd. I guess it works in _this_ case because `_hasInputConnection` is always false when the widget is inflated for the first time? Is that guaranteed?

Reply:
Also, why's this assert even relevant?

Reply:
Yeah the assumption is `_hasInputConnection` is always false when inflated. Connections are opened/closed on focus changes and when the state is inflated the connection is always closed and it won't open until the next microtask flush. The assert is there to make sure the assumption is correct.

Reply:
Looking at the EditableTextState code it seems if you give it a focused FocusNode when it mounts it will just pretend it's not focused.

Reply:
> Looking at the EditableTextState code it seems if you give it a focused FocusNode when it mounts it will just pretend it's not focused.

This sounds like a bug?

> Also, why's this assert even relevant?

I still don't understand why the assert is relevant for the functionality and correctness of `_CompositionCallback`. If it is not needed, I would just remove it (because it is fairly confusing). And if it is needed to assert on some condition not really relevant to the `_CompositionCallback` class itself we should at least leave a comment here and explain that relationship to a future reader.